Crash in WebProcessCache::evictAtRandomIfNeeded()

rniwa opened the pull request:
#### cd134386b5e86ffd45a3574f8bfcb4b5e160053b
<pre>
Crash in WebProcessCache::evictAtRandomIfNeeded()
<a href="https://bugs.webkit.org/show_bug.cgi?id=300279">https://bugs.webkit.org/show_bug.cgi?id=300279</a>

Reviewed by Ben Nham.

The bug was caused by WebProcessCache::evictAtRandomIfNeeded always assuming m_sharedProcessesPerSite
to be not empty, which isn&apos;t necessarily the case. Fixed the bug by adding isEmpty check. Also added
the same check for m_processesPerSite for extra robustness.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm

* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::evictAtRandomIfNeeded):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, WebProcessCacheCrashWithZeroSharedProcess)):

Canonical link: <a href="https://commits.webkit.org/301117@main">https://commits.webkit.org/301117@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e74c84b76f8f6d6e583dbecc270d8112ff1b613

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124977 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44646 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35383 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131826 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76854 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4aa79fc7-e578-4128-9bf5-e34617afd0ec) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126854 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45338 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53212 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95118 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/15c38c44-bca3-4600-be62-37edd615548d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127931 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36179 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111771 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75665 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3510301f-7958-483c-b3fb-25fece616ec7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35108 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29924 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75306 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105942 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30155 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134500 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51800 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39600 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103596 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52225 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107984 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103372 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48713 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27002 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48825 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19584 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51684 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57479 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51058 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54415 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52750 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/WebProcessCache.cpp:
```diff
@@ -189,22 +189,24 @@ bool WebProcessCache::addProcess(Ref<CachedProcess>&& cachedProcess)
 {
-    while (size() >= capacity()) {
-        {
+    while (size() >= capacity() && (m_sharedProcessesPerSite.isEmpty() || m_processesPerSite.isEmpty())) {
```
Not sure I understand the logic behind `&& (m_sharedProcessesPerSite.isEmpty() || m_processesPerSite.isEmpty())`. Doesn't that mean that we will never evict if both of the maps are non-empty?

bnham replied:
Maybe it would be easier to define capacity as an inclusive limit, then you could write:

```
while (size() > capacity()) {
   ...
   if (size() <= capacity()) break;
   ...
}
```

rniwa replied:
Oh oops, I need to put ! in front of each isEmpty check lol.

rniwa committed:
Crash in WebProcessCache::evictAtRandomIfNeeded()
https://bugs.webkit.org/show_bug.cgi?id=300279

Reviewed by Ben Nham.

The bug was caused by WebProcessCache::evictAtRandomIfNeeded always assuming m_sharedProcessesPerSite
to be not empty, which isn't necessarily the case. Fixed the bug by adding isEmpty check. Also added
the same check for m_processesPerSite for extra robustness.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm

* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::evictAtRandomIfNeeded):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, WebProcessCacheCrashWithZeroSharedProcess)):

Canonical link: https://commits.webkit.org/301117@main
